pkg/liqo-controller-manager/resourceoffer-controller/resourceoffer_controller_methods.go

```diff
@@ -0,0 +1,113 @@
+package resourceoffercontroller
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	configv1alpha1 "github.com/liqotech/liqo/apis/config/v1alpha1"
+	discoveryv1alpha1 "github.com/liqotech/liqo/apis/discovery/v1alpha1"
+	sharingv1alpha1 "github.com/liqotech/liqo/apis/sharing/v1alpha1"
+	crdclient "github.com/liqotech/liqo/pkg/crdClient"
+	"github.com/liqotech/liqo/pkg/discovery"
+	"github.com/liqotech/liqo/pkg/utils"
+)
+
+// WatchConfiguration watches a ClusterConfig for reconciling updates on ClusterConfig.
+func (r *ResourceOfferReconciler) WatchConfiguration(kubeconfigPath string, localCrdClient *crdclient.CRDClient, wg *sync.WaitGroup) {
+	defer wg.Done()
+	utils.WatchConfiguration(func(configuration *configv1alpha1.ClusterConfig) {
+		r.setConfig(configuration)
+	}, localCrdClient, kubeconfigPath)
+}
+
+func (r *ResourceOfferReconciler) getConfig() *configv1alpha1.ClusterConfig {
+	r.configurationMutex.RLock()
+	defer r.configurationMutex.RUnlock()
+
+	return r.configuration.DeepCopy()
+}
+
+func (r *ResourceOfferReconciler) setConfig(config *configv1alpha1.ClusterConfig) {
+	r.configurationMutex.Lock()
+	defer r.configurationMutex.Unlock()
+
+	if r.configuration == nil {
+		r.configuration = config
+		return
+	}
+	if !reflect.DeepEqual(r.configuration, config) {
+		r.configuration = config
+	}
+}
+
+// setOwnerReference sets owner reference to the related ForeignCluster.
+func (r *ResourceOfferReconciler) setOwnerReference(ctx context.Context, resourceOffer *sharingv1alpha1.ResourceOffer) (update bool, err error) {
+	// get the foreign cluster by clusterID label
+	foreignClusterList := discoveryv1alpha1.ForeignClusterList{}
+	if err = r.Client.List(ctx, &foreignClusterList, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{
+			discovery.ClusterIDLabel: resourceOffer.Spec.ClusterId,
+		}),
+	}); err != nil {
+		klog.Error(err)
+		return false, err
+	}
+
+	if len(foreignClusterList.Items) == 0 {
```

Comment:
To enforce a controller-reference, you can rely on **https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#SetControllerReference**